### PR TITLE
Make co2 non-static

### PR DESCRIPTION
**Pull request by multitalentloes**
Static classes, in particular static buffers are causing some troubles in the GPU porting of OPM. This PR seeks to avoid the usage of a static implementation of the CO2 class.

WIP

### CODE REVIEW

File: opm/material/components/co2tablesStructs.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (atgeirr) [inline opm/material/components/co2tablesStructs.hpp:1]: Missing license and copyright header.

File: opm/material/components/co2tablesStructs.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef OPM_CO2HELPER_STRUCTS_HPP
+#define OPM_CO2HELPER_STRUCTS_HPP
+
+#include <vector>
+struct TabulatedDensityTraits {
```
Comment (atgeirr) [inline opm/material/components/co2tablesStructs.hpp:6]: No namespace?

File: opm/material/components/co2tablesStructs.hpp (hunk 1 above)
Comment (multitalentloes) [reply opm/material/components/co2tablesStructs.hpp:5]: updated.

File: opm/material/components/co2tablesStructs.hpp (hunk 2 above)
Comment (multitalentloes) [reply opm/material/components/co2tablesStructs.hpp:31]: fixed

File: opm/material/components/CO2Parameters.hpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	static constexpr double brineSalinity = 1.000000000000000e-01;
+
+	CO2Tables();
+};
+
+class CO2Parameters
+{
+public:
+    CO2Parameters(){
+      co2Tables_ = CO2Tables();
```
Comment (atgeirr) [inline opm/material/components/CO2Parameters.hpp:48]: Nitpick: bad an d    in con sistentformatting (2-8 spaces indent!), apply clang-format to new files as a rule of thumb.

More importantly: why a separate CO2Parameters class when CO2Tables could be used directly?

Comment (multitalentloes) [reply opm/material/components/CO2Parameters.hpp]: All three new files have run clang-format now

Comment (multitalentloes) [reply opm/material/components/CO2Parameters.hpp]: CO2Parameters now bypassed and using CO2Tables directly

File: opm/material/components/CO2.hpp [hunk 4]
```diff
@@ -247,6 +289,19 @@ class CO2 : public Component<Scalar, CO2<Scalar>>
         return (mu0 + dmu)/1.0e6; // conversion to [Pa s]
     }
 
+    /*!
+     * \brief The dynamic viscosity [Pa s] of CO2.
+     *
+     * This is only here so that the API is valid
```
Comment (atgeirr) [inline opm/material/components/CO2.hpp:295]: Lots of these functions added, can we not just remove them? What would be the consequence?

File: opm/material/fluidsystems/blackoilpvt/Co2GasPvt.cpp [hunk 5]
```diff
@@ -50,17 +50,19 @@ Co2GasPvt(const std::vector<Scalar>& salinity,
     setActivityModelSalt(activityModel);
     setThermalMixingModel(thermalMixingModel);
 
+    co2Tables = Params();
```
Comment (atgeirr) [inline opm/material/fluidsystems/blackoilpvt/Co2GasPvt.cpp:53]: Will not this just do the entire table building over again, since it already would be default-initialized at this point?

Also in BrineCo2Pvt.cpp.

File: opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.hpp [hunk 6]
```diff
@@ -489,6 +491,16 @@ class BrineCo2Pvt
         return convertXoGToRs(convertxoGToXoG(xlCO2, salinity), regionIdx);
     }
 
+    /*!
+     * \brief Get the CO2 parameters.
+     *
+     * \return A const reference to the CO2 parameters.
+     */
+    const Params& getCo2Tables() const
```
Comment (atgeirr) [inline opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.hpp:499]: Is this just for debugging? Can it be removed? If not, perhaps also add it to the Co2GasPvt class.

File: opm/material/fluidsystems/BrineCO2FluidSystem.hpp [hunk 7]
```diff
@@ -51,10 +52,10 @@ namespace Opm {
 
 // Silence compiler warnings about use of variables
 // that are instantiated in a different compilation unit.
-template<>
-const float CO2<float>::brineSalinity;
-template<>
-const double CO2<double>::brineSalinity;
+// template<>
```
Comment (atgeirr) [inline opm/material/fluidsystems/BrineCO2FluidSystem.hpp:55]: Why is this removed?

File: opm/material/fluidsystems/BrineCO2FluidSystem.hpp [hunk 8]
```diff
@@ -305,7 +306,8 @@ class BrineCO2FluidSystem
         }
 
         assert(phaseIdx == gasPhaseIdx);
-        LhsEval result = CO2::gasViscosity(temperature, pressure);
+        CO2Tables params; // TODO: avoid this initialization
```
Comment (atgeirr) [inline opm/material/fluidsystems/BrineCO2FluidSystem.hpp:309]: I agree with the TODO...

We must find a solution to this (all of the similar TODO places here) before it can be merged.

**Review by atgeirr — COMMENTED** [review]
A very good start, we need to discuss how to access the tables in the BrineCO2FluidSystem though.

File: opm/material/components/CO2.hpp (hunk 4 above)
Comment (multitalentloes) [reply opm/material/components/CO2.hpp:295]: removing these functions causes compilation errors, seems related to the common interface described in "Components.hpp", althought I am not sure why it does not just inherit that implementation which just throws an error

Comment (multitalentloes) [reply opm/material/components/CO2.hpp:295]: And having added new versions of these that also take in the parameter objects it is a bit ugly as inheriting the interface from components does not make it obvious that you are supposed to create a CO2Tables object and send that in instead, not sure how to clean this up nicely

File: opm/material/fluidsystems/BrineCO2FluidSystem.hpp (hunk 7 above)
Comment (multitalentloes) [reply opm/material/fluidsystems/BrineCO2FluidSystem.hpp:55]: I have just set the value inside the CO2 class, I dont think these lines are needed now

File: opm/material/components/CO2Tables.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+#include "CO2tables.inc"
```
[the remark this answers is not in the document]
Comment (multitalentloes) [reply opm/material/components/CO2Tables.cpp]: fixed

File: opm/material/fluidsystems/BrineCO2FluidSystem.hpp (hunk 7 above)
Comment (multitalentloes) [reply opm/material/fluidsystems/BrineCO2FluidSystem.hpp:55]: I'll readd them iff the warnings are still there in jenkins

**Comment by multitalentloes** [comment]
Works with new tables generated by https://github.com/OPM/opm-common/pull/4260 in opm-utilities

File: opm/material/fluidsystems/BrineCO2FluidSystem.hpp (hunk 8 above)
Comment (multitalentloes) [reply opm/material/fluidsystems/BrineCO2FluidSystem.hpp]: Resolved with a static singleton. This will not cause problems for the GPUification of the code as Flow does not use this fluid system regardless.

**Comment by multitalentloes** [comment]
jenkins build this please

File: opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.cpp [hunk 10]
```diff
@@ -94,7 +97,9 @@ initFromState(const EclipseState& eclState, const Schedule&)
     // set the surface conditions using the STCOND keyword
     Scalar T_ref = eclState.getTableManager().stCond().temperature;
     Scalar P_ref = eclState.getTableManager().stCond().pressure;
-    
+
+    co2Tables_ = Params();
```
Comment (atgeirr) [inline opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.cpp:101]: This is double work, as `co2Tables_` should already be default-initialized in the constructor.

File: opm/material/fluidsystems/blackoilpvt/Co2GasPvt.cpp (hunk 5 above)
Comment (atgeirr) [reply opm/material/fluidsystems/blackoilpvt/Co2GasPvt.cpp:53]: Still applies.

**Review by atgeirr — CHANGES_REQUESTED** [review]
This is good to go, except for the double initialization issue.

File: opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.cpp (hunk 10 above)
Comment (multitalentloes) [reply opm/material/fluidsystems/blackoilpvt/BrineCo2Pvt.cpp:99]: fixed now

**Comment by multitalentloes** [comment]
jenkins build this please

**Review by atgeirr — APPROVED** [review]
